File: src/lib/api/importaciones.ts
import { invoke } from "./client";

export type EstadoImportacion = "en_revision" | "confirmada" | "descartada";

export type Importacion = {
  id: number;
  archivoNombre: string;
  archivoHash: string;
  hojaDetectada: string;
  filaEncabezadoDetectada: number;
  columnasDetectadasJson: string;
  estado: EstadoImportacion;
  totalFilas: number;
  creadaEn: string;
  cerradaEn: string | null;
};

export type ClasificacionFila =
  | "producto_valido"
  | "seccion"
  | "ignorada"
  | "requiere_revision"
  | "error";

export type DecisionFila = "crear_nuevo" | "vincular_existente" | "omitir";

export type ImportacionFila = {
  id: number;
  importacionId: number;
  filaExcel: number;
  codigoExcel: string | null;
  nombreExcel: string | null;
  /** Centavos. */
  precioListaCentavos: number | null;
  categoriaExcelTexto: string | null;
  preciosCalculadosJson: string | null;
  clasificacion: ClasificacionFila;
  motivoError: string | null;
  esDuplicadoCodigo: boolean;
  esPosibleDuplicadoNombre: boolean;
  coincideProductoExistenteId: number | null;
  decision: DecisionFila | null;
  productoVinculadoId: number | null;
  actualizarCostoEnVinculo: boolean;
  productoId: number | null;
  resueltaEn: string | null;
};

export type ResumenImportacion = {
  total: number;
  validos: number;
  sinCodigo: number;
  sinNombre: number;
  duplicadosCodigo: number;
  duplicadosNombre: number;
  coincideExistente: number;
  secciones: number;
  ignoradas: number;
  errores: number;
  pendientes: number;
  resueltas: number;
};

export type ColumnaDetectada = {
  encabezado: string;
  indice: number;
};

export type ColumnasDetectadas = {
  codigo: ColumnaDetectada | null;
  nombre: ColumnaDetectada;
  precio: ColumnaDetectada;
  derivadas: ColumnaDetectada[];
};

export type ResolverFila = {
  decision: DecisionFila;
  productoVinculadoId: number | null;
  actualizarCostoEnVinculo: boolean;
  nombreCorregido: string | null;
  codigoCorregido: string | null;
};

export function procesarArchivo(ruta: string, archivoNombre: string) {
  return invoke<Importacion>("importaciones_procesar_archivo", { ruta, archivoNombre });
}

export function listarImportaciones() {
  return invoke<Importacion[]>("importaciones_listar");
}

export function obtenerImportacion(id: number) {
  return invoke<Importacion>("importaciones_obtener", { id });
}

export function obtenerResumen(id: number) {
  return invoke<ResumenImportacion>("importaciones_resumen", { id });
}

export function listarFilas(id: number) {
  return invoke<ImportacionFila[]>("importaciones_listar_filas", { id });
}

export function buscarConfirmadaConMismoHash(hash: string, excluyendoId: number) {
  return invoke<Importacion | null>("importaciones_buscar_confirmada_con_mismo_hash", {
    hash,
    excluyendoId,
  });
}

export function resolverFila(filaId: number, datos: ResolverFila) {
  return invoke<ImportacionFila>("importaciones_resolver_fila", { filaId, datos });
}

export function aplicarPendientes(id: number) {
  return invoke<number>("importaciones_aplicar_pendientes", { id });
}

export function descartarImportacion(id: number) {
  return invoke<Importacion>("importaciones_descartar", { id });
}

export function parsearColumnasDetectadas(json: string): ColumnasDetectadas {
  return JSON.parse(json) as ColumnasDetectadas;
}

File: src-tauri/src/commands/importaciones.rs
use tauri::State;

use crate::db::AppState;
use crate::error::AppResult;
use crate::models::importacion::{Importacion, ImportacionFila, ResolverFila, ResumenImportacion};
use crate::services::importaciones;

#[tauri::command]
pub async fn importaciones_procesar_archivo(
    state: State<'_, AppState>,
    ruta: String,
    archivo_nombre: String,
) -> AppResult<Importacion> {
    importaciones::procesar_archivo(&state.pool, std::path::Path::new(&ruta), archivo_nombre).await
}

#[tauri::command]
pub async fn importaciones_listar(state: State<'_, AppState>) -> AppResult<Vec<Importacion>> {
    importaciones::listar(&state.pool).await
}

#[tauri::command]
pub async fn importaciones_obtener(state: State<'_, AppState>, id: i64) -> AppResult<Importacion> {
    importaciones::obtener(&state.pool, id).await
}

#[tauri::command]
pub async fn importaciones_resumen(
    state: State<'_, AppState>,
    id: i64,
) -> AppResult<ResumenImportacion> {
    importaciones::resumen(&state.pool, id).await
}

#[tauri::command]
pub async fn importaciones_listar_filas(
    state: State<'_, AppState>,
    id: i64,
) -> AppResult<Vec<ImportacionFila>> {
    importaciones::listar_filas(&state.pool, id).await
}

#[tauri::command]
pub async fn importaciones_buscar_confirmada_con_mismo_hash(
    state: State<'_, AppState>,
    hash: String,
    excluyendo_id: i64,
) -> AppResult<Option<Importacion>> {
    importaciones::buscar_confirmada_con_mismo_hash(&state.pool, &hash, excluyendo_id).await
}

#[tauri::command]
pub async fn importaciones_resolver_fila(
    state: State<'_, AppState>,
    fila_id: i64,
    datos: ResolverFila,
) -> AppResult<ImportacionFila> {
    importaciones::resolver_fila(&state.pool, fila_id, datos).await
}

#[tauri::command]
pub async fn importaciones_aplicar_pendientes(
    state: State<'_, AppState>,
    id: i64,
) -> AppResult<i64> {
    importaciones::aplicar_pendientes_con_decision(&state.pool, id).await
}

#[tauri::command]
pub async fn importaciones_descartar(
    state: State<'_, AppState>,
    id: i64,
) -> AppResult<Importacion> {
    importaciones::descartar(&state.pool, id).await
}
